### ISSUE TITLE
Thinking about a combined "share" GUI

### ISSUE THREAD

**Issue by psh** [commons-app/apps-android-commons#1128]
I spent some time in [Sketch](https://www.sketchapp.com/) today thinking about what a user interface might look like if we combined the single-share and multi-share activities into a single workflow.  The thought experiment was about sharing 4 images with the Commons app, noting that we've had requests for 

* the ability to add title + description to all images in a set
* pinch + zoom to scale the image
* ability to review images and see them more clearly
* clearer presentation of the policy declaration that doesn't interfere with the image

![share activity - step 1](https://user-images.githubusercontent.com/407647/36012777-10ae0470-0d26-11e8-984b-c5ce44178a40.png)

Both the top and bottom cards would be (independently) collapsable to allow full access to the image.  Without the previous smoky transparent overlay the image is fully visible.  The small images in the top card scroll horizontally and tapping one pulls it up as the full screen image for review.

![share activity - step 1a](https://user-images.githubusercontent.com/407647/36012817-5b94d2b6-0d26-11e8-8eae-904d08af5c2f.png)![share activity collapsed - step 1a](https://user-images.githubusercontent.com/407647/36012837-7e32544c-0d26-11e8-80ac-7458d6414e73.png)

Note: if you were uploading only a single image then the top card would be hidden, and this first step of the wizard would be skipped taking you directly to adding title and description.

![share activity - step 2](https://user-images.githubusercontent.com/407647/36012870-b7afd55a-0d26-11e8-8212-b769f3284355.png)

The wizard would step forward through the images making each the background of the whole screen to let you pick the best title + description.  We could name them the same as we do now - prefill the title - with the set name and a sequence number.  It would make sense to auto-collapse the top card as the wizard is now in control of which image is "selected" in the preview list - its tied directly to the step of the wizard that we are on.

![stack](https://user-images.githubusercontent.com/407647/36013068-ad7404b6-0d27-11e8-8b23-018669f66c95.png)

When we're done stepping through giving titles to them all, we add categories.  As with the current upload these are common across all the group.  Is there a need to tie them directly to each image?

![share activity - step 6](https://user-images.githubusercontent.com/407647/36013159-36babd00-0d28-11e8-9424-efc4467e0945.png)

Lastly we would select the license and show the _"By submitting this picture, I declare that this is my own work, that it does not contain copyrighted material or selfies, and otherwise adheres to Wikimedia Commons policies."_ at the point of tapping submit.

![share activity - step 7](https://user-images.githubusercontent.com/407647/36013179-649e542a-0d28-11e8-8550-345bf6f2b86e.png)

The number of steps is dynamic based on the number of images shared.

If I share a single image then we would show a 3 step wizard

* Step 1 of 3 - enter title + description
* Step 2 of 3 - select categories
* Step 3 of 3 - license and policy click-through to *submit*

**Comment by nicolas-raoul** [comment]
Thanks for these nice ideas!

One thing: I am very hesitant to reduce the space available for category search. Category search is the most important part of the app, I would say, as uncategorized images are not worth their bytes. Finding the right categories is hard enough with full-screen on a large smartphone (I scroll hundreds of category suggestions for each picture I upload: location-based suggestions, history-based suggestions, filename-based suggestions, then by trying various words/synonyms describing depicted things), I can not imagine what it would be with only 3 lines :-/

**Comment by misaochan** [comment]
I love these! I think it's a step in the right direction in general, especially the collapsible top and bottom cards.

A few concerns:
- Is there a need for "name this set" for multiple uploads? We do have that currently (when you select multiple images in gallery and tap 'share'), but it actually REPLACES the title/desc, it isn't additional to it. This is legacy code, which we are likely to ditch in favour of a uniform title/desc screen for all.
- I think the category selection screen could be automatically uncollapsed, since by that stage it is more important to have the screen real-estate for categories, not the image. This might help alleviate the issue @nicolas-raoul raised
- Categories must be the last step (after license etc) if we want to maintain our current flow of initiating image upload before the user selects categories. We cannot submit a license after the image has already been uploaded, as AFAIK licenses are mandatory for all uploads.

**Comment by psh** [comment]
So, tweaking thoughts about the category search - is this more of what you were thinking?

![share activity - step 6 - alternate](https://user-images.githubusercontent.com/407647/36045381-270c54e0-0d9b-11e8-8e53-4753376c8b61.png)

allowing for the top card to (optionally) be collapsed / expanded to see more

![share activity - step 6 - alternate collapsed](https://user-images.githubusercontent.com/407647/36045399-36dedd20-0d9b-11e8-83a3-de42eb8344d5.png)

**Comment by psh** [comment]
It would be amazing if there was a server component to the upload process - the client sends a single multipart form submission containing both the image & it's categories. From the app perspective it's one seamless transaction that either works or it doesn't - nothing gets out of step - while on the server it can be processed in as many steps as needed.

**Comment by psh** [comment]
I lightly tweaked the final step of the wizard - added a white background and the state (expanded / collapsed) of the top card would carry over from step 6 - to save the "Step __ of __" jumping around as you tap "Next".

![share activity - step 7](https://user-images.githubusercontent.com/407647/36069951-596a235a-0eb7-11e8-9445-3170e8b4601a.png) ![share activity - step 7 - collapsed](https://user-images.githubusercontent.com/407647/36069953-5e3d33ae-0eb7-11e8-9384-c0d9599e2add.png)

And to get a sense of the expected behavior when the keyboard is opened - space between the cards would collapse as far as possible, and then the screen would scroll depending on collapsed/expanded state of cards, screen size, etc.

![share activity - step 5 with keyboard](https://user-images.githubusercontent.com/407647/36070066-0f21e042-0eb9-11e8-8464-c8f120bd6765.png)

**Comment by sivaraam** [comment]
Just a little nitpick about the Licensing step. In case my assumption that, the text "Wikimedia Commons" is a link to the actual policy in Wikimedia Commons, is correct. Then I guess the link text should actually be "Wikimedia Commons policies".

**Comment by psh** [comment]
Thanks for the feedback @sivaraam - I've updated the [design mockup](https://scene.zeplin.io/project/59f8f32ff01c1ed40c7bcee7) to include your thoughts.

**Comment by ilgazer** [comment]
Is there any interest in working on this feature?

**Comment by misaochan** [comment]
@ilgazer Yes, I am thinking about including this as a task in our future PG application. :) Or are you interested in working on it? It will be quite a big task, though, as it would involve a substantial overhaul of the underlying mechanism as well, not just the UI. I think @psh has done part of it at #1503 , but the PR isn't functional yet at the moment.

If we implement this we also need to think about how to handle existing features in ShareActivity like the Zoom and Coordinates buttons, as well as information about licenses etc.

**Comment by ilgazer** [comment]
Unfortunately, I don't have enough free time to undertake such a
significant task, but I believe working towards overhauling the upload
process and finally fully decoupling it from the upload UI should be much
more achievable and could be one of the long-ish term goals of this
project. Unfortunately, I lack the experience in software architecture that
would be required to lead such a project, but I would definitely try to
contribute if you laid out the framework for what needs to be done.

**Comment by ilgazer** [comment]
I took a better look at psh's work in #1503 and it seems that he has done more of the work than I initially remembered. I think I have enough time before school starts to get it to a working state.

**Comment by misaochan** [comment]
@ilgazer That would be wonderful if possible! Please let us know how you get on. Even if you don't have time to do everything, just getting the PR into a functional state would be great. We can then polish it up and add the extra upload components.

**Comment by ilgazer** [comment]
I had lots of free time yesterday and today so I got the PR into a working state. The duplicate and darkness checking etc isn't here yet but uploading works and the category addition workflow is much better(I just append the file) than the "upload and add later" workflow for ShareActivity. As I don't have permission to create new branches on commons-app, could you create an "upload-overhaul" branch so that I can submit a PR to it? IMO getting the initial PR accepted with minimal features would be ideal so that testing can start as soon as possible.

**Comment by misaochan** [comment]
@ilgazer Wonderful! Done, please submit your PR to https://github.com/commons-app/apps-android-commons/tree/upload-overhaul :)

**Comment by ilgazer** [comment]
By the way, should I do separate PR's for new features(darkness checking, some new screens for EXIF editing etc.) as per the "one PR per feature" or go on with this single PR which will end up resolving like 6 issues?

**Comment by misaochan** [comment]
I'm personally OK with either, since it is an "overhaul" PR that will be sent to a separate branch, and all of those changes are related to the overhaul. Thanks for the PR btw, we will test it ASAP. :)

**Comment by misaochan** [comment]
This is almost complete and we will have a brand new upload UI soon. :)

Thank you so much to everyone who contributed to lay down the foundation for @maskaravivek to work on it, especially @psh and @ilgazer . Hugely appreciated!
